### ISSUE TITLE
output of dense_motion 

### ISSUE THREAD

**Issue by instant-high** [yoyo-nb/Thin-Plate-Spline-Motion-Model#21]
Maybe stupid question.

Is it possible to get the deformed/'animated' source_ image before using the inpaint module?

What is the output of the dense motion module.?....

out_dict['contribution_maps'] = contribution_maps
out_dict['deformation'] = deformation # Optical Flow
out_dict['occlusion_map'] = occlusion_map # Multi-resolution Occlusion Masks

out_dict['deformed_source'] = deformed_source ?

Thanks in advance

**Comment by yoyo-nb** [comment]
`out_dict['contribution_maps']` represents the contribution region of each TPS to optical flow

something like this:
![image](https://user-images.githubusercontent.com/48593306/175497428-d9f15e90-baf2-450c-a88a-8288036e2fc5.png)

`out_dict['deformation']` is the optical flow
`out_dict['occlusion_map']` is the Multi-resolution Occlusion Masks mentioned in the paper

`out_dict['deformed_source']` is the deformed source_ image before using the inpaint module

**Comment by instant-high** [comment]
Ok. I had tried that but trying to convert it to display or write it into video file I've got error message. Something like "out_dict has no deformed source image" if I remember correctly.
If it was not that thing, how to transpose?

Maybe you can help me?
I would like to try if my face swap mod is still working then but waý faster.
Using face parsing for masking I get ~ 2.it/sec now

**Comment by yoyo-nb** [comment]
sorry i said wrong

`out_dict['deformed_source']` is **not** the deformed source_ image before using the inpaint module, it's just the source_image deformed by each TPS. 

What you need is [here](https://github.com/yoyo-nb/Thin-Plate-Spline-Motion-Model/blob/main/modules/inpainting_network.py#L102)

**Comment by instant-high** [comment]
Yes, that is working. 
So the deformed source image is not output of the dense motion script? I mean, I can not skip the whole inpaint module?
Line 91 in demo.py
out = inpainting_network(source, dense_motion)

Sorry for asking again

**Comment by yoyo-nb** [comment]
you can try this:
```python
def deform_input(inp, deformation):
    _, h_old, w_old, _ = deformation.shape
    _, _, h, w = inp.shape
    if h_old != h or w_old != w:
        deformation = deformation.permute(0, 3, 1, 2)
        deformation = F.interpolate(deformation, size=(h, w), mode='bilinear', align_corners=True)
        deformation = deformation.permute(0, 2, 3, 1)
    return F.grid_sample(inp, deformation,align_corners=True)

# ............
# demo.py, Line 88
dense_motion = dense_motion_network(source_image=source, kp_driving=kp_norm,
                                                    kp_source=kp_source, bg_param = None, 
                                                    dropout_flag = False)
deformation = dense_motion['deformation']
deformed_source = deform_input(source, deformation)  # deformed source_ image before using the inpaint module
```

**Comment by instant-high** [comment]
that is working so far but giving some strange/funny results. But faster....

I think I keep my hands off from changing this part.
The results of your original demo code are very good.

Thanks a lot
